# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

<<<<<<< HEAD
#' Compute CG content
#' @param x StringMatrix with bases, of reads x cycles
NULL

cpp_GC_content <- function(x) {
    .Call('_FastqCleaner_cpp_GC_content', PACKAGE = 'FastqCleaner', x)
}

=======
>>>>>>> upstream/master
#'Compute the base content in a DNA matrix
#'@param x matrix with nucleotides for each position (column)
#'of the reads (rows)
#'@param relative. Logical. If true (default), the function
#' computes relative frequencies
NULL

cpp_base_content <- function(x, relative = TRUE) {
<<<<<<< HEAD
    .Call('_FastqCleaner_cpp_base_content', PACKAGE = 'FastqCleaner', x, relative)
=======
    .Call('_FastqCleaner_cpp_base_content', 
    PACKAGE = 'FastqCleaner', x, relative)
>>>>>>> upstream/master
}

#' check by row if the values of a matrix are above a 
#' threshold of values in a vector returns a matrix of rows by threshold values
#' @param x NumericMatrix to be checked
#' @param thresVector vector with thresholds
NULL

cpp_check_quality_threshold <- function(x, thresVector) {
<<<<<<< HEAD
    .Call('_FastqCleaner_cpp_check_quality_threshold', PACKAGE = 'FastqCleaner', x, thresVector)
=======
    .Call('_FastqCleaner_cpp_check_quality_threshold', 
    PACKAGE = 'FastqCleaner', x, thresVector)
>>>>>>> upstream/master
}

#' create a StringVector using the size of a given NumericVector
#' @param vec Input NumericVector
#' @param x character used to fill the output vector
NULL

cpp_create_stringvec <- function(vec, x) {
<<<<<<< HEAD
    .Call('_FastqCleaner_cpp_create_stringvec', PACKAGE = 'FastqCleaner', vec, x)
=======
    .Call('_FastqCleaner_cpp_create_stringvec', 
    PACKAGE = 'FastqCleaner', vec, x)
}

#' Compute CG content
#' @param x StringMatrix with bases, of reads x cycles
NULL

cpp_GC_content <- function(x) {
    .Call('_FastqCleaner_cpp_GC_content', 
    PACKAGE = 'FastqCleaner', x)
>>>>>>> upstream/master
}

#' Select the correct match position for matching operations with
#' multiple strings 
#' @description The program takes as input a logical matrix that 
#' indicates with TRUE or FALSE if a match is present in a given position
#'  or not, respectively. Each column is a sequence, each row a cycle
#'  for NGS experiments
#' @param m LogicalMatrix with TRUE and FALSE values
#' @param width length of each sequence. This allow to discard fake matches
#' @param origin If origin = "start", select the first match. 
#' If origin = "end", selects the last match.
#' If origin  = "end_inverted" selects the first match starting 
#' from n_i positions, where n is the  length of the sequence represented 
#' in the column i.  This case works when the last match is 
#' seeked in reversed sequences.
NULL

cpp_which_true <- function(m, width, origin) {
<<<<<<< HEAD
    .Call('_FastqCleaner_cpp_which_true', PACKAGE = 'FastqCleaner', m, width, origin)
=======
    .Call('_FastqCleaner_cpp_which_true', 
    PACKAGE = 'FastqCleaner', m, width, origin)
>>>>>>> upstream/master
}

